### PR TITLE
fix(tar): enable unix paths in tar RootPath

### PR DESCRIPTION
**Pull request by sensslen**
Fix Path RootPath property usage in TarArchive by applying the same transformation as used when stripping a TarEntry path.

see #581 #338 

_I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License._

### CODE REVIEW

**Comment by Numpsy** [comment]
this PR seems to include a set of Zip changes from one of the other (unmerged) async related PRs, rather than being against the master branch or being just the described issue?

**Comment by sensslen** [comment]
You are correct - I got confused by that branch that was called something like origin/.../master. Sorry! I corrected this mistake.

**Comment by piksel** [comment]
Yeah, that's what the `gh` tool does when you check out PRs. I should rename it.

**Comment by sensslen** [comment]
Is there anything witholding this PR to be merged? We're curently spinning our own version of SharpZipLib which includes this fix but I would really like to get rid of it as soon as possible!

**Comment by piksel** [comment]
Well, there are some problems with DropPathRoot. It will throw exceptions on .NET 4.6 if you give it certain paths which alters the behaviour (and can throw upon reading files etc, which isn't great).
See #593 which deals with this. I think I have found all the problems with it though, so that should be merged pretty soon. The plan is to use it for all the path handling code that is currently split between zip and tar. It would basically just slot in as a replacement for the `ClearTarPath()` method in this PR.

**Comment by sensslen** [comment]
Good to hear, that you are working on this issue. Looking forward to the next release

File: src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System.IO;
+
+namespace ICSharpCode.SharpZipLib.Tar
+{
+	internal static class TarStringExtension
+	{
+		public static string ClearTarPath(this string s)
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs:7]: It's not obvious what `ClearTarPath` would do...
```suggestion
		public static string ToTarArchivePath(this string s)
```
This still doesn't explain *how* it's done, but it explains the purpose and what the expected outcome is.

File: src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System.IO;
+
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs:2]:
```suggestion
using ICSharpCode.SharpZipLib.Core;

```

File: src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+	internal static class TarStringExtension
+	{
+		public static string ClearTarPath(this string s)
+		{
+			var pathRoot = Path.GetPathRoot(s);
+			if (!string.IsNullOrEmpty(pathRoot))
+			{
+				s = s.Substring(pathRoot.Length);
+			}
+			return s.Replace(Path.DirectorySeparatorChar, '/');
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarStringExtension.cs:14]:
```suggestion
			return PathUtils.DropPathRoot(s).Replace(Path.DirectorySeparatorChar, '/');
```

File: test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs [hunk 4]
```diff
@@ -899,5 +899,77 @@ public void StreamWithJapaneseName(int length, string encodingName)
...
+			CreateAndClearDirectory(tempDirectory);
+			using (var dummyfile = File.Create(Path.Combine(tempDirectory, "dummyfile"))) 
+			{
+				using (var randomStream = new ChaosStream())
+				{
+					randomStream.CopyTo(dummyfile);
+				}
+			}
+
+			var tarFileName = Path.Combine(Path.GetTempPath(), "sharpziplib_tar_test_folder_archive.tar");
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs:922]:
```suggestion
			using var tempDirectory = new Utils.TempDir();
			tempDirectory.CreateDummyFile();

			using var tarFileName = new Utils.TempFile();
```

File: test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs [hunk 5]
```diff
@@ -899,5 +899,77 @@ public void StreamWithJapaneseName(int length, string encodingName)
...
+				using (var tarOutputStream = TarArchive.CreateOutputTarArchive(tarFile))
+				{
+					tarOutputStream.RootPath = tempDirectory;
+					var entry = TarEntry.CreateEntryFromFile(tempDirectory);
+					tarOutputStream.WriteEntry(entry, true);
+				}
+			}
+
+			var extractDirectory = Path.Combine(Path.GetTempPath(), "sharpziplib_tar_extract_folder");
+			CreateAndClearDirectory(extractDirectory);
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs:935]:
```suggestion
			using var extractDirectory = new Utils.TempDir();
```

File: test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs [hunk 6]
```diff
@@ -899,5 +899,77 @@ public void StreamWithJapaneseName(int length, string encodingName)
...
+			private readonly int length = new Random().Next() % 500000 + 200;
+
+			// Create constructors as needed to match desired MemoryStream construction
+
+			public override int Read(byte[] buffer, int offset, int count)
+			{
+				int readCount = Math.Max(0, Math.Min(length - offset, count));
+				return base.Read(buffer, offset, readCount);
+			}
+		}
```
Comment (piksel) [inline test/ICSharpCode.SharpZipLib.Tests/Tar/TarTests.cs:973]: test/ICSharpCode.SharpZipLib.Tests/TestSupport/Utils.cs contains helpers that does this already (and cleans up after tests, regardless of outcome)
```suggestion
```

File: src/ICSharpCode.SharpZipLib/Tar/TarEntry.cs [hunk 7]
```diff
@@ -418,15 +418,10 @@ public void GetFileTarHeader(TarHeader header, string file)
...
-			name = name.Replace(Path.DirectorySeparatorChar, '/');
-
 			// No absolute pathnames
 			// Windows (and Posix?) paths can start with UNC style "\\NetworkDrive\",
 			// so we loop on starting /'s.
-			while (name.StartsWith("/", StringComparison.Ordinal))
-			{
-				name = name.Substring(1);
-			}
+			name = name.ClearTarPath();
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarEntry.cs:424]:
```suggestion
			name = name.ToTarArchivePath();
```

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs [hunk 8]
```diff
@@ -356,8 +356,7 @@ public string RootPath
 				{
 					throw new ObjectDisposedException("TarArchive");
 				}
-				// Convert to forward slashes for matching. Trim trailing / for correct final path
-				rootPath = value.Replace('\\', '/').TrimEnd('/');
+				rootPath = value.ClearTarPath().TrimEnd('/');
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs:359]:
```suggestion
				rootPath = value.ToTarArchivePath().TrimEnd('/');
```

**Review by piksel — COMMENTED** [review]
Added suggestions (that can be batch-commited if you want) for using the shared `DropPathRoot` and reuse of existing test helpers.

File: src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs (hunk 8 above)
Comment (sensslen) [reply src/ICSharpCode.SharpZipLib/Tar/TarArchive.cs:359]: Thanks a lot for your suggestions. They do absolutely make sense. Just out of curiosity. Didn't #593 address  this issue already and this PR is thus obsolete now?

**Comment by sensslen** [comment]
Thanks a lot for your suggestions. They do absolutely make sense. Just out of curiosity. Didn't #593 address this issue already and this PR is thus obsolete now?

**Comment by sensslen** [comment]
I'd love to hear why this PR is still not merged....

**Comment by piksel** [comment]
> I'd love to hear why this PR is still not merged....

I have been busy with life and other projects. It's still in the pipeline for 1.4.

**Comment by sensslen** [comment]
> > I'd love to hear why this PR is still not merged....
> 
> I have been busy with life and other projects. It's still in the pipeline for 1.4.

No Worries

**Review by piksel — APPROVED** [review]

**Pull request merged**